### PR TITLE
Add basic composite tree model

### PR DESCRIPTION
**Pull request by tlambert03**
# Description
This PR adds a couple (currently unused) classes `Node` and `Group` (a subclass of both `Node` and `NestableEventedList`) that can serve as the basis of a basic composite Tree model.  I have a followup PR that will add Qt Models and Views dedicated to each of these classes.

To implement layer groups (#970), `Layer` would ultimately inherit from `Node`, and `LayerList` would inherit from `Group`.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by tlambert03** [comment]
@sofroniewn, @jni... this is all set, pretty fully tested, and ready for review.  In addition to the new `Node` and `Group` classes, it makes the update to the nestedEventedList class (which we don't really use) that will be necessary for complicated non-contiguous drag-drop in the qt tree model #2222.

This PR is included in that PR if you'd like to see how it all works.

**Comment by sofroniewn** [comment]
This is very exciting! I havn't had a chance to do a deep dive in the code yet, but am wondering how this will interact with making the Layers be `EventedModels` and if that might give us some challenges.

> To implement layer groups (#970), Layer would ultimately inherit from Node, and LayerList would inherit from Group.

Do you think we could do something where instead of Layer inheriting from Node, what we call the Layer right now actually lives as an attribute on the Node? i.e. the tree structure itself is built with nodes and groups, but our `Layers` themselves remain simple `EventedModels`.

One reason for this is I'm worried about the `Layer` becoming very complex - I'm already struggling to use Pydantic with out keybindings/ mouse function mixins (private attributes don't work in mixins right now) and this could help keep concerns separate. I know that it goes against the composite pattern which @jni has been a proponent of, as only leaf nodes would need a `.layer` property but I feel like we never resolved https://github.com/napari/napari/issues/970#issuecomment-624607364 and similar discussions.

I'm also thinking more about your comment here https://github.com/napari/napari/issues/970#issuecomment-772813308

> nope, #970 and #1502 are literal groupings of layers, and not so much "binding" of certain layer attributes. However, it's not so hard to imagine, a relatively simple API for binding the attributes of two multiple layers:

and how nice it is to have that flexibility and how we might want to use that mechanism to do things like couple controlling contrast limits on many layers together or visibility together, instead of using the layer group composite pattern.

One downside is that this might require more of a major breaking change because syntax might need to look `viewer.layers[0].layer` or something like that which maybes isn't as nice.

We can move this discussion to #970 if you want. It's also not something we need to resolve before this merges as we might want to use this structure regardless, but curious if you have thoughts on it now!!

**Comment by tlambert03** [comment]
> This is very exciting! I havn't had a chance to do a deep dive in the code yet, but am wondering how this will interact with making the Layers be EventedModels and if that might give us some challenges.

Nope, it's designed specifically for evented objects, and the concept that you should be able to listen only to the root level object to spy on events happening anywhere in the tree.  

> Do you think we could do something where instead of Layer inheriting from Node, what we call the Layer right now actually lives as an attribute on the Node? i.e. the tree structure itself is built with nodes and groups, but our Layers themselves remain simple EventedModels... I'm worried about the Layer becoming very complex - I'm already struggling to use Pydantic with out keybindings/ mouse function mixins (private attributes don't work in mixins right now) and this could help keep concerns separate. 

I'd really rather not go there already.  It is of course possible to do something like that, but it would totally break our whole "pure python, it's just a list" ethos.  I know you're struggling with the mixins elsewhere, but I don't think these classes will be as challenging.  The main reason being that the `Node` and `Group` classes are pretty much strictly _behavioral_.  They have _tiny_ `__init__` functions and are pretty much all methods.  To me, "separation of concerns" doesn't demand that we use composition everywhere, but rather that "related logic" be nicely grouped together.  The Node and Group classes here do just that, by grouping the behavior required of a Node.  I think these methods-classes will be easy to inherit (unlike the mixins you're struggling with that attributes) in pydantic classes, but if we run into issues, we can address them at that point.  (my confidence also stems from the fact that this was all working already in #1502 ...)

> how we might want to use that mechanism to do things like couple controlling contrast limits on many layers together or visibility together, instead of using the layer group composite pattern.

Personally, I remain quite excited about the composite pattern.  I feel like it will provide the cleanest "slickest" end-user API for the 90% use case of literal groupings of layers.  It _will_ allow controlling attributes on many layers that are grouped together, exactly _because_ of the composite pattern.  If we need to additionally link attributes between layers that needn't be explicitly grouped, then I think something like I proposed in https://github.com/napari/napari/issues/970#issuecomment-772813308 is what we should try first.

**Comment by sofroniewn** [comment]
> Nope, it's designed specifically for evented objects, and the concept that you should be able to listen only to the root level object to spy on events happening anywhere in the tree.
> ...
> They have tiny __init__ functions and are pretty much all methods ... I think these methods-classes will be easy to inherit (unlike the mixins you're struggling with that attributes) in pydantic classes, but if we run into issues, we can address them at that point.

Ok yes, I was thinking that they should be easier to integrate for that reason too, and I think it's fine to wait, but I'm glad we're thinking about it now a little.

> Personally, I remain quite excited about the composite pattern. I feel like it will provide the cleanest "slickest" end-user API for the 90% use case of literal groupings of layers.

Yes agreed on the niceties of the end-user API, I would not give that up easily either!

> If we need to additionally link attributes between layers that needn't be explicitly grouped, then I think something like I proposed in #970 (comment) is what we should try first.

Yeah, I think this is still interesting, and we can continue to discuss ideas around that too on #970 (I have some thoughts around using "tags" that could allow very flexible grouping from an API perspective that belong on that issue rather than this PR).

I'll get to reviewing the code now. As this isn't actually being wired up in this PR it really is fine to delay on my concerns from above, but I think you also did a good job explaining how you're thinking about them :-)

**Comment by tlambert03** [comment]
> > If we need to additionally link attributes between layers that needn't be explicitly grouped, then I think something like I proposed in #970 (comment) is what we should try first.

> Yeah, I think this is still interesting, and we can continue to discuss ideas around that too on #970 (I have some thoughts around using "tags" that could allow very flexible grouping from an API perspective that belong on that issue rather than this PR).

got a working PR... will submit in a min

File: napari/utils/events/containers/_nested_list.py
Comment (sofroniewn) [inline napari/utils/events/containers/_nested_list.py:27]: I find the word `Maybe` a little confusing as its not quite clear what this thing is - it may be a nested index or it may be something else entirely. In actuality it seems like it is either an Index or a NestedIndex. How about something like `AnyIndex`?

I guess on second read it could be a "MaybeNested" Index which makes a little more sense, but its a little confusing in the function signatures bellow

File: napari/utils/events/containers/_nested_list.py
Comment (sofroniewn) [inline napari/utils/events/containers/_nested_list.py:377]: When I was doing things with the evented list we ended up wanting to make some changes to how moving was done that led us to not need to use a blocker - see #2087. I'm not sure the exact situation applies here, and it is something that we can solve if/when it comes up, but I just wanted to flag it as I was reminded of that PR

File: napari/utils/tree/_tests/test_tree_model.py
Comment (sofroniewn) [inline napari/utils/tree/_tests/test_tree_model.py:83]: These methods are really really nice

File: napari/utils/tree/_tests/test_tree_model.py
Comment (sofroniewn) [inline napari/utils/tree/_tests/test_tree_model.py:114]: Can you add
```
    assert [x.name for x in slc] == ['9', 'g1']
```
Assuming I have that right! And maybe a comment like
```
    # slicing is a form of iterating, and iterating a group just uses its children
```

I had to think a little to know what `slc = tree[::-2]` would do

File: napari/utils/tree/_tests/test_tree_model.py
Comment (sofroniewn) [inline napari/utils/tree/_tests/test_tree_model.py:171]: I love this syntax `tree[1, 2]`. I know this maybe testes elsewhere, but a dedicated test which was something like `test_nested_indexing` with things like `assert tree[1, 2].name == '5'` and maybe some with `Error` for things like trying to access `tree[1, 2, 0]` etc might round out this file nicely

Comment (sofroniewn) [reply napari/utils/tree/_tests/test_tree_model.py:171]: Also can the tree be indexed by name, so `assert tree['g1', '5'].name == '5'`. Again that behaviour might be tested in other places not on the Tree, but I see the Tree as something we'll want to engage with a lot so it might be nice to have a test for that here too

File: napari/utils/tree/node.py
Comment (sofroniewn) [inline napari/utils/tree/node.py]: The recursivity comes from the call in `Group`, maybe this doc string should just say it returns the singleton name, right?

File: napari/utils/tree/node.py
Comment (sofroniewn) [inline napari/utils/tree/node.py:9]: Should get doc strings before merge

File: napari/utils/tree/group.py
Comment (sofroniewn) [inline napari/utils/tree/group.py]: Looking at this a group is a node that can have children - I'm sort of wondering if most of the time we're just going to want to have groups and not directly use nodes, but we can see how it all shakes out as we go! It is nice to have option just to have Nodes, which are much simpler as they don't inherit the `NestableEventedList`!!

File: napari/utils/tree/group.py
Comment (sofroniewn) [inline napari/utils/tree/group.py:34]: Group should get doc strings before merge

File: setup.cfg
Comment (sofroniewn) [inline setup.cfg:131]: Just curious what is this/ what does it mean?

**Review by sofroniewn — COMMENTED** [review]
Ok @tlambert03 this is looking very nice, I love the syntax in the tree and the tests. Most of the comments I've left are pretty minor/ for my learning benefit. I think we should be able to get this in soon without any major changes. Good stuff!! I'll look at #2222 now

File: napari/utils/events/containers/_nested_list.py
Comment (tlambert03) [reply napari/utils/events/containers/_nested_list.py:27]: I like `AnyIndex` just as much as `Maybe...`

File: napari/utils/events/containers/_nested_list.py
Comment (tlambert03) [reply napari/utils/events/containers/_nested_list.py:377]: Yeah, I remember that.  I'm _hoping_ this will be ok.  This is the thing that took days to figure out.  to keep a very long story somewhat short: there are basically two ways to sync movements between a python model and a Qt Model
1) using [`begin/endMoveRows`](https://doc.qt.io/qt-5/qabstractitemmodel.html#beginMoveRows), (which only supports moving contiguous items)
2) using [`layoutAboutToBeChanged`](https://doc.qt.io/qt-5/qabstractitemmodel.html#layoutAboutToBeChanged) and manually using [`changePersistentIndex`](https://doc.qt.io/qt-5/qabstractitemmodel.html#changePersistentIndex) to update all of the QModel indices.

the former is a "higher level" API that takes care of some super annoying stuff for you, but it doesn't work with non-contiguous movements (which will be common with drag/drop).  The latter is tremendously flexible, but quite complicated, because you not only need to update the moving indices, but all of the other indices in the model that are "affected" by the moving items.

So, after trying to use the lower level API, I backed off and settled on the `beginMoveRows` API... which basically requires me to move these items one at a time, emitting events each time.  Hopefully it won't break whatever we were avoiding in #2087?

File: setup.cfg
Comment (tlambert03) [reply setup.cfg:131]: that's the "ignore this line" command for [coverage.py](https://coverage.readthedocs.io/en/coverage-4.3.3/excluding.html) ... when you don't care if a given line is covered in test coverage

File: napari/utils/tree/_tests/test_tree_model.py
Comment (tlambert03) [reply napari/utils/tree/_tests/test_tree_model.py:114]: sure!  added your test and the comment `slc = tree[::-2]  # take every other item, starting from the end`

File: napari/utils/tree/_tests/test_tree_model.py
Comment (tlambert03) [reply napari/utils/tree/_tests/test_tree_model.py:171]: > but a dedicated test which was something like `test_nested_indexing`

how about that actual test :joy:  (it's in master already)

https://github.com/napari/napari/blob/a91ba740d767feab9f6f3135bcb34489901ebaa5/napari/utils/events/_tests/test_evented_list.py#L234-L235

> but I see the Tree as something we'll want to engage with a lot so it might be nice to have a test for that here too

hmmm, but all of the logic and methods being tested are over in `_nested_list.py`


> Also can the tree be indexed by name, so assert tree['g1', '5'].name == '5'

oh man :joy: ... Yes, you can search nested lists by name [see `test_nested_custom_lookup`](https://github.com/napari/napari/pull/2217/files#diff-25f46b91b3c6443994ba1261120fb13bb2fd9a682300bd96a64d2d20dc763ce5R138) ... but I don't think you can also do it with a tuple like that.  so, you'd have to do `tree['g1']['5']` ... that one would be a bit tough

File: napari/utils/tree/node.py
Comment (tlambert03) [reply napari/utils/tree/node.py:9]: thanks for noting!

File: napari/utils/events/containers/_nested_list.py
Comment (sofroniewn) [reply napari/utils/events/containers/_nested_list.py:377]: Yeah, I can imagine the complexity here. Let's press on, but thanks for those additional links, might come in handy in the future!!

File: napari/utils/tree/_tests/test_tree_model.py
Comment (sofroniewn) [reply napari/utils/tree/_tests/test_tree_model.py:171]: ok yeah, that's fair maybe not need to duplicate test if literally would be identical and no reason to suspect that any further additions to `Group` would break it.

>  so, you'd have to do tree['g1']['5'] ... that one would be a bit tough

ok, makes sense - maybe one day!

File: setup.cfg
Comment (sofroniewn) [reply setup.cfg:131]: thanks

File: napari/utils/tree/group.py
Comment (tlambert03) [reply napari/utils/tree/group.py]: > Looking at this a group is a node that can have children

yep!  that's the core "composite" idea there.  everything is a node.

> It is nice to have option just to have Nodes, which are much simpler as they don't inherit the NestableEventedList!!

Yeah, `Layer` is a just a `Node`, it can't have children.  but `LayerGroup` (i.e. the future `LayerList`) will be a `Group` and therefore a `NestableEventedList` _and_ it is a `Layer/Node` (that's the big change that will allow `LayerGroups` to have properties like opacity and visibility and stuff.

File: napari/utils/events/containers/_nested_list.py
Comment (tlambert03) [reply napari/utils/events/containers/_nested_list.py:27]: Actually... as I started changing this, I realize that `AnyIndex` loses the "`Nested`" part... which is more important.  ok if I keep `Maybe`?

Comment (tlambert03) [reply napari/utils/events/containers/_nested_list.py:27]: (i could also do `PossiblyNestedIndex` if you prefer)

Comment (sofroniewn) [reply napari/utils/events/containers/_nested_list.py:27]: we can leave it as `MaybeNestedIndex` unless @jni  has another idea

**Review by sofroniewn — APPROVED** [review]
Alright, this looks good to me!! Thanks for adding those comments in. As this PR doesn't actually use the tree model anywhere yet I think it's fine to merge now as is. We can leave this open overnight in case @jni wants to weigh in, but @tlambert03 if you'd like to merge this in the morning when you wake up then I say go for it. You can then update #2222 and I can give that a review

**Comment by sofroniewn** [comment]
Ok, I'm merging this now. @jni can look through at a later date. It will be nice to press on with reviewing the followup PRs and this PR doesn't wire anything up yet. Excited to see this moving forward!!

**Pull request merged**
